app/controllers/sessions_controller.rb

```diff
@@ -2,19 +2,46 @@ class SessionsController < ApplicationController
   def new; end
 
   def create
-    user = User.find_by(email: params[:session][:email].downcase)
-    if user&.authenticate(params[:session][:password])
-      reset_session
-      log_in user
-      redirect_to user
+    user = find_user_by_email
+    if user_authenticated?(user)
+      handle_successful_login(user)
     else
-      flash.now[:danger] = t "login.danger_flash"
-      render :new, status: :unprocessable_entity
+      handle_failed_login
     end
   end
 
   def destroy
-    log_out
+    log_out if logged_in?
     redirect_to root_url, status: :see_other
   end
+
+  private
+
+  def find_user_by_email
+    User.find_by(email: params[:session][:email].downcase)
```

Comment:
```suggestion
    @user = User.find_by email: params.dig(:session, :email)&.downcase
    return if @user
    
    flash[:error] = "User not found" # I18n đi nhé
    redirect_to root_path
```
thêm code handle user nil luôn để các method sau sử dụng ko cần kết hợp toán tử an toàn

Reply:
e nghĩ là khác này chỉ có nhiệm vụ tìm kiếm `user` và `return user or nil`. còn việc xử lý báo lỗi và redirect sẽ thuộc nhiệm vụ của method khác. E cũng xử lý báo lỗi ở method `handle_failed_login` rồi ạ